pod/perlhacktips.pod

```diff
@@ -309,6 +309,123 @@ C<-xc99>
+several types.  A function has to be declared with a single explicit
+type.  This can be finessed in some situations by adding the type of the
+parameter as an extra accompanying parameter:
+
+ #define foo(param, type)  STMT_START { type param = 0; ... } STMT_END
+
+Or maybe the code involved is so trivial that a function would be just
+complicating overkill, such as when the macro simply creates a mnemonic
+name for some constant value.
+
+If you do choose to use a non-trivial macro, be aware that there are
+several avoidable pitfalls that can occur.  Keep in mind that a macro is
+expanded within the lexical context of each place in the source it is
+called.  If you have a token C<foo> in the macro and the source happens
+also to have C<foo>, the meaning of the macro's C<foo> will become that
+of the caller's.  Sometimes that is exactly the behavior you want, but
+be aware that this tends to be confusing later on.  It effectively turns
+C<foo> into a reserved word for any code that calls the macro, and this
+fact is usually not documented nor considered.  It is safer to pass
+C<foo> as a parameter, so that C<foo> remains freely available to the
+caller and the macro interface is explicitly specified.
+
+Worse is when the equivalence between the two C<foo>'s is coincidental.
+Suppose for example, that the macro declares a variable
+
+ int foo
+
+That works fine as long as the caller doesn't define the string C<foo>
+in some way.  And it might not be until years later that someone comes
+along with an instance where C<foo> is used.  For example a future
+caller could do this:
+
+ #define foo  bar
+
+Then that declaration of C<foo> in the macro suddenly becomes
+
+ int bar
+
+That could mean that something completely different happens than
+intended.  It is hard to debug; the macro and call may not even be in
+the same file, so it would require some digging and gnashing of teeth to
+figure out.
+
+Therefore, if a macro does use variables, their names should be such
+that it is very unlikely that they would collide with any caller, now or
+forever.  One way to do that, now being used in the perl source, is to
+include the name of the macro itself as part of the name of each
+variable in the macro.  Suppose the macro is named C<SvPV>  Then we
+could have
+
+ int foo_svpv_ = 0;
+
+This is harder to read than plain C<foo>, but it is pretty much
+guaranteed that a caller will never naively use C<foo_svpv_> (and run
+into problems).  (The lowercasing makes it clearer that this is a
+variable, but assumes that there won't be two elements whose names
+differ only in the case of their letters.)  The trailing underscore
+makes it even more unlikely to clash, as those, by convention, signify a
+private variable name.  (Note that C reserves, for its own use, variable
+names containing two underscores in a row, or beginning with an
+underscore.  It's undefined behavior for anyone else to do that.)
```

Comment:
A really common solution in our code is to variables that AlTeRnAtInG case. I have not seen trailing underbars used often but i have seen the alternating case form used a bunch. I think Nick Clark liked the style.

Reply:
I'm hesitant to mention this.  I find those difficult to read myself; having to stop and unfocus my eyes so as to read the whole word in a slurp.  I read once that non-native-English speakers have a hard time with PascalCase in general.

And it doesn't solve the problem of macro X calling macro Y and both use the same variable name.  That requires something unique to the instance.  The solution cited is easier to read than ALtErNaTeCaSe, and works except if X gets called recursively.  Thus it is both easier to read and works in more cases.

I've thought about some way to use the line number in the file of the declare statement as part of the the variable name, like we do when we emulate static assertions.  But haven't come up with a good solution,.